test/dist_autograd_test.py

```diff
@@ -10,15 +10,36 @@
 
 import threading
 
-prev_rank_rpc_done = False
-prev_rank_context_id = 0
+# Right now we test up to 3-layer nested rpc calls.
+# rpc_done[1] and ctx_ids[1] represent rpc is done in prev rank, and context id
+# sent from prev rank respectively.
+# rpc_done[2] and ctx_ids[2] represents for prev of prev rank.
+# rpc_done[3] and ctx_ids[3] represents for prev of prev of prev rank.
+rpc_done = [False, False, False, False]
+ctx_ids = [-1, -1, -1, -1]
 
 
-def _set_rpc_done(context_id):
-    global prev_rank_rpc_done
-    global prev_rank_context_id
-    prev_rank_rpc_done = True
-    prev_rank_context_id = context_id
+# Send rpc done info and context_id to
+# dst_rank = (self.rank + rank_distance) % self.world_size
+def _set_rpc_done(ctx_id, rank_distance):
+    global rpc_done
+    global ctx_ids
+    rpc_done[rank_distance] = True
+    ctx_ids[rank_distance] = ctx_id
+
+
+def my_py_add(t1, t2):
+    return torch.add(t1, t2)
+
+
+def my_py_nested_call(t1, t2, dst, world_size, hops):
+    next_dst = (dst + 1) % world_size
+    if hops > 0:
+        return rpc.rpc_sync("worker{}".format(next_dst), my_py_nested_call,
```

Comment:
(Does not need to be in this PR). Let's also test async rpc calls, e.g., making multiple async calls, collect all futures in a list, and wait on all list in the end.